ports/stm32/machine_i2s.c

```diff
@@ -0,0 +1,1109 @@
+// - the DMA ping-pong buffer needs to be aligned to a cache line size of 32 bytes.  32 byte
+//   alignment is needed to use the routines that clean and invalidate D-Cache which work on a
+//   32 byte address boundary.  Not all STM32 devices have a D-Cache.  Buffer alignment
+//   will still happen on these devices to keep this code simple.
+
+#define MAX_I2S_STM32 (2)
+
+// DMA ping-pong buffer size was empirically determined.  It is a tradeoff between:
+// 1. memory use (smaller buffer size desirable to reduce memory footprint)
+// 2. interrupt frequency (larger buffer size desirable to reduce interrupt frequency)
+// The sizeof 1/2 of the DMA buffer must be evenly divisible by the cache line size of 32 bytes.
+#define SIZEOF_DMA_BUFFER_IN_BYTES (256)
+#define SIZEOF_HALF_DMA_BUFFER_IN_BYTES (SIZEOF_DMA_BUFFER_IN_BYTES / 2)
+
+// For non-blocking mode, to avoid underflow/overflow, sample data is written/read to/from the ring buffer at a rate faster
+// than the DMA transfer rate
+#define NON_BLOCKING_RATE_MULTIPLIER (4)
+#define SIZEOF_NON_BLOCKING_COPY_IN_BYTES (SIZEOF_HALF_DMA_BUFFER_IN_BYTES * NON_BLOCKING_RATE_MULTIPLIER)
+
+#define NUM_I2S_USER_FORMATS (4)
+#define I2S_RX_FRAME_SIZE_IN_BYTES (8)
+
+typedef enum {
+    MONO,
+    STEREO
+} format_t;
+
+typedef enum {
+    BLOCKING,
+    NON_BLOCKING,
+    UASYNCIO
+} io_mode_t;
+
+typedef enum {
+    TOP_HALF,
+    BOTTOM_HALF
+} ping_pong_t;
+
+typedef struct _ring_buf_t {
+    uint8_t *buffer;
+    size_t head;
+    size_t tail;
+    size_t size;
+} ring_buf_t;
+
+typedef struct _non_blocking_descriptor_t {
+    mp_buffer_info_t appbuf;
+    uint32_t index;
+    bool copy_in_progress;
+} non_blocking_descriptor_t;
+
+typedef struct _machine_i2s_obj_t {
+    mp_obj_base_t base;
+    uint8_t i2s_id;
+    mp_hal_pin_obj_t sck;
+    mp_hal_pin_obj_t ws;
+    mp_hal_pin_obj_t sd;
+    uint16_t mode;
+    int8_t bits;
+    format_t format;
+    int32_t rate;
+    int32_t bufferlen;
+    mp_obj_t callback_for_non_blocking;
+    uint8_t dma_buffer[SIZEOF_DMA_BUFFER_IN_BYTES + 0x1f]; // 0x1f related to D-Cache alignment
+    uint8_t *dma_buffer_dcache_aligned;
+    ring_buf_t ring_buffer;
+    uint8_t *ring_buffer_storage;
+    non_blocking_descriptor_t non_blocking_descriptor;
+    io_mode_t io_mode;
+    I2S_HandleTypeDef hi2s;
+    DMA_HandleTypeDef hdma_tx;
+    DMA_HandleTypeDef hdma_rx;
+    const dma_descr_t *dma_descr_tx;
+    const dma_descr_t *dma_descr_rx;
+} machine_i2s_obj_t;
+
+// The frame map is used with the readinto() method to transform the audio sample data coming
+// from DMA memory (32-bit stereo) to the format specified
+// in the I2S constructor.  e.g.  16-bit mono
+STATIC const int8_t i2s_frame_map[NUM_I2S_USER_FORMATS][I2S_RX_FRAME_SIZE_IN_BYTES] = {
+    { 0,  1, -1, -1, -1, -1, -1, -1 },  // Mono, 16-bits
+    { 2,  3,  0,  1, -1, -1, -1, -1 },  // Mono, 32-bits
+    { 0,  1,  4,  5, -1, -1, -1, -1 },  // Stereo, 16-bits
```

Comment:
I think this code assumes that Mono always corresponds to LEFT - it's possible that an I2S device could be set up to be on the RIGHT channel.